WebContent logs during launch are not emitted

pvollan opened the pull request:
#### 2bf1c7444f37c96f1d786793e7c555e5d8e4bac4
<pre>
WebContent logs during launch are not emitted
<a href="https://bugs.webkit.org/show_bug.cgi?id=291839">https://bugs.webkit.org/show_bug.cgi?id=291839</a>
<a href="https://rdar.apple.com/149683032">rdar://149683032</a>

Reviewed by Chris Dumez.

WebContent logs during launch are not emitted when logd is blocked in the sandbox because the log streaming channel
to the UI process is not set up until WebProcess::platformInitializeWebProcess is called. This patch addresses this
by sending logs emitted in this time interval over the XPC connection to the UI process and emit the logs there.

Test: Tools/TestWebKitAPI/Tests/WebKit/LogForwarding.mm

* Source/JavaScriptCore/API/JSValue.mm:
(createStructHandlerMap):
* Source/JavaScriptCore/API/JSWrapperMap.mm:
(parsePropertyAttributes):
* Source/JavaScriptCore/API/ObjcRuntimeExtras.h:
(adoptSystem): Deleted.
* Source/WTF/wtf/SystemFree.h:
(WTF::adoptSystem):
* Source/WebKit/Shared/Cocoa/LaunchLogHook.h: Added.
* Source/WebKit/Shared/Cocoa/LaunchLogHook.mm: Added.
(WebKit::LaunchLogHook::singleton):
(WebKit::LaunchLogHook::initialize):
(WebKit::LaunchLogHook::disable):
* Source/WebKit/Shared/Cocoa/LaunchLogMessages.h: Added.
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::initializeLogd):
(WebKit::XPCServiceEventHandler):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _launchLogsCountForTesting]):
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::xpcEventHandler const):
(WebKit::WebProcessProxy::XPCEventHandler::handleXPCEvent const):
(WebKit::WebProcessProxy::XPCEventHandler::XPCEventHandler):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::registerLogClient):
* Tools/TestWebKitAPI/Tests/WebKit/LogForwarding.mm:
(TEST(WebKit, LaunchLogs)):

Canonical link: <a href="https://commits.webkit.org/300971@main">https://commits.webkit.org/300971@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/315e81ecba94966bc1a2da55f082ff72e523c28f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124448 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44131 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34863 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131292 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76459 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/1ec7cd58-4617-4a70-b606-1c7f44e72813) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126325 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44871 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52726 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94682 "16 failures") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/62802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2599787d-22f4-4dd1-b865-7953c94ace69) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127402 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35769 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111314 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75259 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ee42c024-182d-4d2f-8be3-c646fd8d74a1) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34705 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29475 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74770 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116581 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105521 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29698 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133957 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/122978 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51345 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39172 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103157 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51746 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107533 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102946 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48315 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26569 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48290 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19541 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51211 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56998 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/154073 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50628 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39146 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53990 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52305 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Shared/Cocoa/LaunchLogHook.mm:
```diff
@@ -0,0 +1,89 @@
+        if (type == OS_LOG_TYPE_FAULT)
+            type = OS_LOG_TYPE_ERROR;
+
+        if (char* messageString = os_log_copy_message_string(msg)) {
```
We should really avoid explicit `free()` in new code, I recommend including `<wtf/SystemFree.h>` and using:
```
if (std::unique_ptr<const char[], SystemFree<const char[]>> messageString = os_log_copy_message_string(msg)) {
```

cdumez replied:
Bonus point if we can move `adoptSystem()` from `Source/JavaScriptCore/API/ObjcRuntimeExtras.h` to `wtf/SystemFree.h` use use this here:
```
if (auto messageString = adoptSystem(os_log_copy_message_string(msg)) {
```

I think `adoptSystem()` naming is not great but at least it is convenient to use. Maybe I can come up with a better name and follow-up.

cdumez commented on Source/WebKit/Shared/Cocoa/LaunchLogHook.mm:
```diff
@@ -0,0 +1,89 @@
+            xpc_dictionary_set_string(message.get(), messageStringKey, messageString);
+            xpc_dictionary_set_uint64(message.get(), logTypeKey, type);
+            xpc_connection_send_message(LaunchLogHook::singleton().m_connection.get(), message.get());
+            free(messageString);
```
Noooo :)

cdumez commented on Source/WebKit/Shared/Cocoa/LaunchLogMessages.h:
```diff
@@ -0,0 +1,43 @@
+
+namespace WebKit {
+
+constexpr ASCIILiteral logMessageName = "log-message"_s;
```
Presumably could use `auto` instead of `ASCIILiteral`.

cdumez commented on Source/WebKit/Shared/Cocoa/LaunchLogReceiver.h:
```diff
@@ -0,0 +1,53 @@
+private:
+    LaunchLogReceiver() = default;
+
+    bool m_enabled { true };
```
Missing prefix per WebKit coding style: `m_isEnabled`

cdumez commented on Source/WebKit/Shared/Cocoa/LaunchLogReceiver.mm:
```diff
@@ -0,0 +1,72 @@
+
+        auto subsystem = xpcDictionaryGetString(event, subsystemKey);
+        auto category = xpcDictionaryGetString(event, categoryKey);
+        auto message_string = xpcDictionaryGetString(event, messageStringKey);
```
variable names should be camel case.

cdumez commented on Source/WebKit/Shared/Cocoa/LaunchLogReceiver.mm:
```diff
@@ -0,0 +1,72 @@
+        auto logType = xpc_dictionary_get_uint64(event, logTypeKey);
+        auto pid = xpc_connection_get_pid(xpc_dictionary_get_remote_connection(event));
+
+        auto osLog = OSObjectPtr<os_log_t>();
```
It's a weird way to write:
```
OSObjectPtr<os_log_t> osLog;
```

cdumez commented on Source/WebKit/Shared/Cocoa/LaunchLogReceiver.mm:
```diff
@@ -0,0 +1,72 @@
+        if (!subsystem.isEmpty() && !category.isEmpty())
+            osLog = adoptOSObject(os_log_create(subsystem.utf8().data(), category.utf8().data()));
+
+        auto osLogPointer = osLog.get() ? osLog.get() : OS_LOG_DEFAULT;
```
Wouldn't `osLog ? ` work? without the `.get()`

cdumez commented on Source/WebKit/Shared/Cocoa/LaunchLogReceiver.mm:
```diff
@@ -0,0 +1,72 @@
+
+        auto osLogPointer = osLog.get() ? osLog.get() : OS_LOG_DEFAULT;
+        os_log_with_type(osLogPointer, static_cast<os_log_type_t>(logType), "WebContent[%d] %{public}s", static_cast<int>(pid), message_string.utf8().data());
+        m_logCountForTesting++;
```
`++ m_logCountForTesting;` always good practice to use pre-increment whenever possible.

cdumez commented on Source/WebKit/Shared/Cocoa/LaunchLogHook.mm:
```diff
@@ -0,0 +1,89 @@
+
+LaunchLogHook& LaunchLogHook::singleton()
+{
+    static NeverDestroyed<LaunchLogHook> logHook;
```
If this can get called from multiple threads (which I suspect it can), we should really be using `LazyInitialize` and `std::call_once()`.

cdumez commented on Source/WebKit/Shared/Cocoa/LaunchLogReceiver.mm:
```diff
@@ -0,0 +1,72 @@
+
+LaunchLogReceiver& LaunchLogReceiver::singleton()
+{
+    static NeverDestroyed<LaunchLogReceiver> logReceiver;
```
If this can get called from multiple threads (which I suspect it can), we should really be using `LazyInitialize` and `std::call_once()`.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKit/LogForwarding.mm:
```diff
@@ -90,4 +90,10 @@
 
+TEST(WebKit, LaunchLogs)
+{
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:NSMakeRect(0, 0, 300, 300)]);
```
`auto` -> `RetainPtr` in new code.

cdumez commented on Source/WebKit/Shared/Cocoa/LaunchLogReceiver.mm:
```diff
@@ -0,0 +1,72 @@
+        os_log_with_type(osLogPointer, static_cast<os_log_type_t>(logType), "WebContent[%d] %{public}s", static_cast<int>(pid), message_string.utf8().data());
+        m_logCountForTesting++;
+    } else if (messageName == disableLogMessageName) {
+        m_enabled = false;
```
There is something fishy here or I am missing something. We have a single receiver in the UIProcess since this is a singleton. We also have many senders potentially since each WebContent process has its own sender (also singleton per process). But then it seems like one of these senders can send a "disable" message to the singleton in the UIProcess and disable it for ALL web processes? This doesn't seem right.

I don't even really understand why we need to disable the receiver in the UIProcess, if the WebProcesses stop sending after launch is complete, then it seems we should not need to ignore any messages on UIProcess side and since this is a singleton, we're not even freeing anymore on disablement.

pvollan replied:
Ah, good catch! Will fix.

Thanks for reviewing!

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.h:
```diff
@@ -878,6 +887,18 @@ class WebProcessProxy final : public AuxiliaryProcessProxy {
     HashMap<String, SandboxExtension::Handle> m_fileSandboxExtensions;
+
+#if PLATFORM(COCOA)
+    class XPCEventHandler : public WebKit::XPCEventHandler {
```
Can be marked as `final`. Do we need `WebKit::` here really?

pvollan replied:
I had to keep `WebKit::` to disambiguate between `WebKit::WebProcessProxy::XPCEventHandler` and `WebKit::XPCEventHandler`.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.h:
```diff
@@ -878,6 +887,18 @@ class WebProcessProxy final : public AuxiliaryProcessProxy {
+#if PLATFORM(COCOA)
+    class XPCEventHandler : public WebKit::XPCEventHandler {
+    public:
+        XPCEventHandler(const WebProcessProxy&);
```
Can this be `explicit`?

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.h:
```diff
@@ -878,6 +887,18 @@ class WebProcessProxy final : public AuxiliaryProcessProxy {
+    public:
+        XPCEventHandler(const WebProcessProxy&);
+
+        bool handleXPCEvent(xpc_object_t) const override;
```
Can be marked as `final`.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
```diff
@@ -367,6 +373,53 @@
+        auto osLogPointer = osLog ? osLog.get() : OS_LOG_DEFAULT;
+        os_log_with_type(osLogPointer, static_cast<os_log_type_t>(logType), "WebContent[%d] %{public}s", static_cast<int>(pid), messageString.utf8().data());
+        ++webProcess->m_logCountForTesting;
+    } else if (messageName == disableLogMessageName) {
```
As in my previous review, I still do not understand why we need the WebProcess to send a `disableLog` message to the UIProcess. I don't think I got an answer to my question? Why can't the WebProcess just stop sending messages (it seems like it always does)?

pvollan replied:
That's a good point. The reasoning behind this is to avoid processing unexpected messages on the XPC channel after the WebContent process has stopped sending logs. I will add a message check in the upcoming check instead of the early return.

Thanks for reviewing!

cdumez replied:
Is it really worth the extra complexity? Don't we anyway destroy both the xpc_connection and this object once the process is done launching?

pvollan replied:
That is a good point! We destroy the XPC connection on iOS, so this is only relevant on macOS. I can conditionalize on that feature :)

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
```diff
@@ -367,6 +373,53 @@
+        RefPtr webProcess = m_webProcess.get();
+        if (!webProcess)
+            return true;
+        webProcess->m_logEndpointEnabled = false;
```
Why does this need to be stored on WebProcessProxy? Why can't it simply be a data member of this class? Seems like it would be better not to pollute WebProcessProxy if we could avoid it.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
```diff
@@ -367,6 +373,53 @@
+
+        auto osLogPointer = osLog ? osLog.get() : OS_LOG_DEFAULT;
+        os_log_with_type(osLogPointer, static_cast<os_log_type_t>(logType), "WebContent[%d] %{public}s", static_cast<int>(pid), messageString.utf8().data());
+        ++webProcess->m_logCountForTesting;
```
Can we make this name less generic? maybe `m_logCountDuringLaunchForTesting`? Also, given how this is used, maybe this can simply be a boolean: `bool m_didReceiveLogsDuringLaunchForTesting`

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
```diff
@@ -1076,6 +1076,15 @@ - (unsigned)_forwardedLogsCountForTesting
+- (unsigned)_launchLogsCountForTesting
+{
+#if ENABLE(LOGD_BLOCKING_IN_WEBCONTENT)
+    return _page->protectedLegacyMainFrameProcess()->logCountForTesting();
```
Can you avoid calling `protectedLegacyMainFrameProcess()` in new code? We should probably be getting the process from the main frame instead.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
```diff
@@ -367,6 +373,56 @@
+        m_didReceiveLogsDuringLaunchForTesting = true;
+        return true;
+    }
+#if !ENABLE(CLOSE_WEBCONTENT_XPC_CONNECTION_POST_LAUNCH)
```
I was trying to get rid of this logic to reduce complexity, not increase complexity by adding build flag checks. I would drop them. They seem incomplete anyway since `m_logEndpointEnabled` is not guarded.

pvollan replied:
Good point! I removed these in the latest patch.

Thanks for reviewing!

cdumez reviewed (approved):
LGTM

pvollan committed:
WebContent logs during launch are not emitted
https://bugs.webkit.org/show_bug.cgi?id=291839
rdar://149683032

Reviewed by Chris Dumez.

WebContent logs during launch are not emitted when logd is blocked in the sandbox because the log streaming channel
to the UI process is not set up until WebProcess::platformInitializeWebProcess is called. This patch addresses this
by sending logs emitted in this time interval over the XPC connection to the UI process and emit the logs there.

Test: Tools/TestWebKitAPI/Tests/WebKit/LogForwarding.mm

* Source/JavaScriptCore/API/JSValue.mm:
(createStructHandlerMap):
* Source/JavaScriptCore/API/JSWrapperMap.mm:
(parsePropertyAttributes):
* Source/JavaScriptCore/API/ObjcRuntimeExtras.h:
(adoptSystem): Deleted.
* Source/WTF/wtf/SystemFree.h:
(WTF::adoptSystem):
* Source/WebKit/Shared/Cocoa/LaunchLogHook.h: Added.
* Source/WebKit/Shared/Cocoa/LaunchLogHook.mm: Added.
(WebKit::LaunchLogHook::singleton):
(WebKit::LaunchLogHook::initialize):
(WebKit::LaunchLogHook::disable):
* Source/WebKit/Shared/Cocoa/LaunchLogMessages.h: Added.
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::initializeLogd):
(WebKit::XPCServiceEventHandler):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _launchLogsCountForTesting]):
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::xpcEventHandler const):
(WebKit::WebProcessProxy::XPCEventHandler::handleXPCEvent const):
(WebKit::WebProcessProxy::XPCEventHandler::XPCEventHandler):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::registerLogClient):
* Tools/TestWebKitAPI/Tests/WebKit/LogForwarding.mm:
(TEST(WebKit, LaunchLogs)):

Canonical link: https://commits.webkit.org/300971@main
